### PR TITLE
feat: implement Snowflake-compatible substr function

### LINKED ISSUE #1311: Create our version of substr
Current datafusion version of `substr` and `substring` is compatible with PostgreSQL. The main difference is treatment of negative numbers in second argument. For example:
`SELECT substr('mystring', -1, 3);` -> result `m` in PostgreSQL and datafusion
`SELECT substr('mystring', -1, 3);` - > result `g` in Snowflake.
We should create our own version of `substr` which follow this documentation: 
https://docs.snowflake.com/en/sql-reference/functions/substr .
To complete this task we need:
1. Follow guidelines in crates/embucket-functions/docs/function_implementation_guide.md 
2. Use the datafusion substr implementation as reference
3. Create snapshot tests in embucket-functions/src/tests/string_binary folder.

### PR DESCRIPTION
**Pull request by Vedin**
# feat: implement Snowflake-compatible substr function

## Summary

This PR implements a Snowflake-compatible `substr` function to replace the current PostgreSQL-compatible version. The key difference is in how negative start positions are handled:

- **PostgreSQL/DataFusion behavior**: `substr('mystring', -1, 3)` → `"mys"` (negative positions clamped to 0)
- **Snowflake behavior**: `substr('mystring', -1, 3)` → `"g"` (negative positions count from end of string)

**Changes made:**
- Added new `SubstrFunc` struct implementing `ScalarUDFImpl` trait
- Implemented `compute_snowflake_substr` function with Snowflake-style negative index logic
- Added support for both `substr` and `substring` aliases
- Created comprehensive snapshot tests covering edge cases, Unicode, and NULL handling
- Registered function in `string-binary/mod.rs`

**Core logic**: For negative start positions, `actual_start = char_count + start + 1`, so `substr('mystring', -1, 3)` becomes `substr('mystring', 8, 3)` starting from position 8 (the last character).

## Review & Testing Checklist for Human

**⚠️ Critical (4 items):**
- [x] **Verify the implementation compiles and tests pass** - Due to edition2024 environment issues, I couldn't run cargo test/fmt/clippy
- [x] **Test the specific issue examples**: `SELECT substr('mystring', -1, 3)` should return `'g'` (not `'mys'`)
- [x] **Verify both aliases work**: Test both `substr(...)` and `substring(...)` functions
- [x] **Run the new snapshot tests**: Ensure all 32 test cases in `src/tests/string_binary/substr.rs` pass

**Recommended test plan:**
1. Run `cargo test -p embucket-functions string_binary::substr` to verify all tests pass
2. Test the issue examples in a SQL session: `SELECT substr('mystring', -1, 3), substr('mystring', -3, 2)`
3. Verify no regressions in existing string functions
4. Check that the function is discoverable in SQL queries

---

### Diagram

```mermaid
%%{ init : { "theme" : "default" }}%%
graph TD
    Issue["Issue #1311<br/>Snowflake substr compatibility"]
    
    subgraph "New Implementation"
        SubstrRS["src/string-binary/substr.rs<br/>SubstrFunc + compute_snowflake_substr"]:::major-edit
        SubstrTests["src/tests/string_binary/substr.rs<br/>32 snapshot tests"]:::major-edit
    end
    
    subgraph "Integration Points"
        ModRS["src/string-binary/mod.rs<br/>Function registration"]:::minor-edit
        TestModRS["src/tests/string_binary/mod.rs<br/>Test module registration"]:::minor-edit
    end
    
    subgraph "Reference Implementation"
        DatafusionSubstr["datafusion/functions/src/unicode/substr.rs<br/>PostgreSQL-compatible logic"]:::context
    end
    
    Issue --> SubstrRS
    SubstrRS --> ModRS
    SubstrTests --> TestModRS
    DatafusionSubstr -.-> SubstrRS
    
    subgraph Legend
        L1["Major Edit"]:::major-edit
        L2["Minor Edit"]:::minor-edit
        L3["Context/No Edit"]:::context
    end
    
    classDef major-edit fill:#90EE90
    classDef minor-edit fill:#87CEEB
    classDef context fill:#FFFFFF
```

### Notes

- **Environment limitation**: Could not run tests locally due to edition2024 requiring newer Cargo version
- **Manual verification**: Logic was verified manually against Snowflake documentation examples
- **Function tracking**: The `cargo update-functions` command failed, so function tracking may need manual update
- **Implementation follows**: embucket-functions implementation guide patterns from existing functions like `lower.rs`

**Link to Devin run**: https://app.devin.ai/sessions/d502d94137444ced8d0ea60ddbad8e0c  
**Requested by**: @Vedin

Fixes #1311

### CODE REVIEW

File: crates/embucket-functions/src/tests/string_binary/substr.rs [hunk 1]
```diff
@@ -0,0 +1,182 @@
+use crate::tests::macros::test_query;
+
+test_query!(
```
Comment (osipovartem) [inline crates/embucket-functions/src/tests/string_binary/substr.rs:3]: snapshots are missing? :)

File: crates/embucket-functions/src/string-binary/substr.rs [hunk 2]
```diff
@@ -0,0 +1,482 @@
...
+    fn name(&self) -> &str {
+        "substr"
+    }
+
+    fn signature(&self) -> &Signature {
+        &self.signature
+    }
+
+    fn return_type(&self, _arg_types: &[DataType]) -> DFResult<DataType> {
+        Ok(DataType::Utf8View)
```
Comment (Vedin) [inline crates/embucket-functions/src/string-binary/substr.rs]: Change it to UTF8

File: crates/embucket-functions/src/tests/string_binary/substr.rs (hunk 1 above)
Comment (Vedin) [reply crates/embucket-functions/src/tests/string_binary/substr.rs:3]: `cargo insta test -- embucket-functions` should be run to complete this and verify this tests

File: crates/embucket-functions/src/string-binary/substr.rs [hunk 3]
```diff
@@ -0,0 +1,532 @@
...
+        Ok(DataType::Utf8)
+    }
+
+    fn aliases(&self) -> &[String] {
+        &self.aliases
+    }
+
+    fn coerce_types(&self, arg_types: &[DataType]) -> DFResult<Vec<DataType>> {
+        if arg_types.len() < 2 || arg_types.len() > 3 {
+            return plan_err!(
```
Comment (rampage644) [inline crates/embucket-functions/src/string-binary/substr.rs]: Errors

File: crates/embucket-functions/src/tests/string_binary/substr.rs (hunk 1 above)
Comment (Vedin) [reply crates/embucket-functions/src/tests/string_binary/substr.rs:3]: done

File: crates/embucket-functions/src/string-binary/substr.rs (hunk 2 above)
Comment (Vedin) [reply crates/embucket-functions/src/string-binary/substr.rs]: done

File: crates/embucket-functions/src/string-binary/substr.rs (hunk 3 above)
Comment (Vedin) [reply crates/embucket-functions/src/string-binary/substr.rs]: done

**Review by osipovartem — APPROVED** [review]

File: crates/embucket-functions/src/string-binary/errors.rs [hunk 4]
```diff
@@ -31,6 +31,80 @@ pub enum Error {
         #[snafu(implicit)]
         location: Location,
     },
+
+    #[snafu(display(
+        "The {function_name} function requires {expected} arguments, but got {actual}"
+    ))]
+    InvalidArgumentCount {
```
Comment (rampage644) [inline crates/embucket-functions/src/string-binary/errors.rs:38]: nitpick: Perhaps we want such errors to be defined even at a higher level since those are applicable to many functions

**Review by rampage644 — APPROVED** [review]

File: crates/embucket-functions/src/string-binary/errors.rs (hunk 4 above)
Comment (Vedin) [reply crates/embucket-functions/src/string-binary/errors.rs:38]: Ok, I'll do it next time working on such functions
